File: Python/Problem_050.py
from math import isqrt

def sieve(limit: int) -> bytearray:
    is_prime = bytearray(b"\x01") * (limit + 1)
    is_prime[0:2] = b"\x00\x00"
    p = 2
    while p * p <= limit:
        if is_prime[p]:
            start = p * p
            is_prime[start::p] = b"\x00" * (((limit - start) // p) + 1)
        p += 1
    return is_prime

def solve(limit: int = 1_000_000) -> int:
    is_prime = sieve(limit)
    primes = [i for i in range(2, limit) if is_prime[i]]

    pref = [0]
    s = 0
    for p in primes:
        s += p
        pref.append(s)

    best_len = 0
    best_prime = 0

    n = len(primes)
    for i in range(n):
        # smallest possible sum for length best_len+1 starting at i
        for j in range(i + best_len + 1, n + 1):
            val = pref[j] - pref[i]
            if val >= limit:
                break
            if is_prime[val]:
                best_len = j - i
                best_prime = val

    return best_prime

if __name__ == "__main__":
    print(solve())  # 997651

File: CPP/Problem_050.cpp
#include <iostream>
#include <vector>
using namespace std;

static vector<char> sieve(int limit) {
    vector<char> isPrime(limit + 1, 1);
    isPrime[0] = isPrime[1] = 0;
    for (int p = 2; 1LL * p * p <= limit; ++p) {
        if (isPrime[p]) {
            for (long long x = 1LL * p * p; x <= limit; x += p)
                isPrime[(int)x] = 0;
        }
    }
    return isPrime;
}

int main() {
    const int LIMIT = 1000000;
    auto isPrime = sieve(LIMIT);

    vector<int> primes;
    for (int i = 2; i < LIMIT; ++i) {
        if (isPrime[i]) primes.push_back(i);
    }

    vector<long long> pref(primes.size() + 1, 0);
    for (size_t i = 0; i < primes.size(); ++i) {
        pref[i + 1] = pref[i] + primes[i];
    }

    int bestLen = 0;
    int bestPrime = 0;

    int n = (int)primes.size();
    for (int i = 0; i < n; ++i) {
        for (int j = i + bestLen + 1; j <= n; ++j) {
            long long sum = pref[j] - pref[i];
            if (sum >= LIMIT) break;
            if (isPrime[(int)sum]) {
                bestLen = j - i;
                bestPrime = (int)sum;
            }
        }
    }

    cout << bestPrime << "\n"; // 997651
    return 0;
}
